payjoin/src/uri.rs

```diff
@@ -1,12 +1,109 @@
 use std::borrow::Cow;
 use std::convert::TryFrom;
 
-use bitcoin::address::{Error, NetworkChecked, NetworkUnchecked};
-use bitcoin::Network;
+use bip21::{DeserializationError, DeserializeParams};
+use bitcoin::address::{Error, NetworkChecked, NetworkUnchecked, NetworkValidation};
+use bitcoin::{Address, Amount, Network};
 use url::Url;
 
-/// Payjoin enum represents payjoin parameters in
-/// different states.
+/// Payjoin Uri represents a bip21 uri with additional
+/// payjoin parameters.
+pub struct PayjoinUri<
+    'a,
+    N: NetworkValidation,
+    P: DeserializeParams<'a> + DeserializationError + Sized,
+> {
+    inner: bip21::Uri<'a, N, P>,
+}
+
+impl<'a, N: NetworkValidation, P: DeserializeParams<'a> + DeserializationError + Sized>
+    From<bip21::Uri<'a, N, P>> for PayjoinUri<'a, N, P>
+{
+    fn from(value: bip21::Uri<'a, N, P>) -> Self { Self { inner: value } }
+}
+
+impl<'a> PayjoinUri<'a, NetworkUnchecked, Payjoin> {
+    /// Marks network of this address as checked.
+    pub fn assume_checked(self) -> PayjoinUri<'a, NetworkChecked, Payjoin> {
+        let checked_address = self.inner.address.assume_checked();
+        let mut uri = bip21::Uri::with_extras(checked_address, self.inner.extras);
```

Comment:
We should be using `PayjoinUri::with_extras` so we know that we access `with_extras` only from a single place